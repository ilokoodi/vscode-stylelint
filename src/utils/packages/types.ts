import { Stylelint } from '../stylelint/index';

/**
 * Package manager identifiers.
 */
export type PackageManager = 'npm' | 'yarn' | 'pnpm';

/**
 * Options for resolving the Stylelint package.
 */
export type ResolverOptions = {
	packageManager?: PackageManager;
	stylelintPath?: string;
};

/**
 * Stylelint package resolution result.
 */
export type StylelintResolutionResult = {
	stylelint: Stylelint;
	resolvedPath: string;
};

/**
 * A tracer function that can be used to log messages.
 */
export type TracerFn = (message: string, verbose?: string) => void;
