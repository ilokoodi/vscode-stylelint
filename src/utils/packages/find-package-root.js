'use strict';

const fs = require('fs/promises');
const path = require('path');

/**
 * Walks up the file tree from the given directory until it finds a directory
 * containing a file named `package.json`. Resolves to `undefined` if no such
 * directory is found.
 * @param {string} directory The directory to start from.
 * @returns {Promise<string | undefined>}
 */
async function findPackageRoot(directory) {
	let currentDirectory = directory;

	// eslint-disable-next-line no-constant-condition
	while (true) {
		const manifestPath = path.join(currentDirectory, 'package.json');

		try {
			const stat = await fs.stat(manifestPath);

			if (stat.isFile()) {
				return currentDirectory;
			}

			const parent = path.dirname(currentDirectory);

			if (!path.relative(parent, currentDirectory)) {
				return undefined;
			}

			currentDirectory = parent;
		} catch (error) {
			if (/** @type {{code?: string}} */ (error).code === 'ENOENT') {
				const parent = path.dirname(currentDirectory);

				if (!path.relative(parent, currentDirectory)) {
					return undefined;
				}

				currentDirectory = parent;
			} else {
				throw error;
			}
		}
	}
}

module.exports = {
	findPackageRoot,
};
