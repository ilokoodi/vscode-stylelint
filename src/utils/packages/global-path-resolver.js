'use strict';

const os = require('os');
const path = require('path');
const { runProcessFindLine } = require('../processes');

/** @type {{[key in PackageManager]: (trace?: TracerFn, isWindows?: boolean) => Promise<string | undefined>}} */
const resolvers = {
	/**
	 * Resolves the global `node_modules` path for Yarn.
	 *
	 * Note: Only Yarn 1.x is supported. Yarn 2.x and higher have removed
	 * support for globally installed packages.
	 */
	async yarn(trace, isWindows) {
		const tryTrace = trace ?? (() => undefined);
		/** @param {string} line */
		const tryParseLog = (line) => {
			/** @type {{type: any, data: any}} */
			let log;

			try {
				log = JSON.parse(line);
			} catch {
				return undefined;
			}

			return log;
		};

		const yarnGlobalPath = await runProcessFindLine(
			'yarn',
			['global', 'dir', '--json'],
			isWindows ? { shell: true } : undefined,
			(line) => {
				const log = tryParseLog(line);

				if (!log || log.type !== 'log' || !log.data) {
					return undefined;
				}

				const globalPath = path.join(log.data, 'node_modules');

				tryTrace(`Yarn returned global path: "${globalPath}"`);

				return globalPath;
			},
		);

		if (!yarnGlobalPath) {
			tryTrace('"yarn global dir --json" did not return a path.');

			return undefined;
		}

		return yarnGlobalPath;
	},

	/**
	 * Resolves the global `node_modules` path for npm.
	 */
	async npm(trace, isWindows) {
		const tryTrace = trace ?? (() => undefined);
		const npmGlobalPath = await runProcessFindLine(
			'npm',
			['config', 'get', 'prefix'],
			isWindows ? { shell: true } : undefined,
			(line) => {
				const trimmed = line.trim();

				if (!trimmed) {
					return undefined;
				}

				const globalPath =
					os.platform() === 'win32'
						? path.join(trimmed, 'node_modules')
						: path.join(trimmed, 'lib/node_modules');

				tryTrace(`npm returned global path: "${globalPath}"`);

				return globalPath;
			},
		);

		if (!npmGlobalPath) {
			tryTrace('"npm config get prefix" did not return a path.');

			return undefined;
		}

		return npmGlobalPath;
	},

	/**
	 * Resolves the global `node_modules` path for pnpm.
	 */
	async pnpm(trace, isWindows) {
		const tryTrace = trace ?? (() => undefined);
		const pnpmGlobalPath = await runProcessFindLine(
			'pnpm',
			['root', '-g'],
			isWindows ? { shell: true } : undefined,
			(line) => {
				const trimmed = line.trim();

				if (!trimmed) {
					return undefined;
				}

				tryTrace(`pnpm returned global path: "${trimmed}"`);

				return trimmed;
			},
		);

		if (!pnpmGlobalPath) {
			tryTrace('"pnpm root -g" did not return a path.');

			return undefined;
		}

		return pnpmGlobalPath;
	},
};

/**
 * Returns an object with a `resolve` method that takes as its first parameter
 * a supported package manager (`npm`, `yarn`, or `pnpm`) and as its second an
 * optional tracer function that will be used to trace the resolution process.
 *
 * On a successful resolution, the method returns a promise that resolves to the
 * package manager's global `node_modules` path. Paths are cached in the
 * resolver on the first successful resolution.
 *
 * When a path cannot be resolved, the promise resolves to `undefined`.
 *
 * @example
 * ```js
 * const resolver = getGlobalPathResolver();
 * const yarnGlobalPath = await resolver.resolve(
 *   'yarn',
 *   message => connection && connection.tracer.log(message)
 * );
 * ```
 * @returns {GlobalPathResolver}
 */
function getGlobalPathResolver() {
	/** @type {GlobalPathResolverCache} */
	const cache = {};

	return {
		async resolve(packageManager, trace) {
			const cached = cache[packageManager];

			if (cached) {
				return cached;
			}

			const tryTrace = trace ?? (() => undefined);
			const resolver = resolvers[packageManager];

			if (!resolver) {
				tryTrace(`Package manager "${packageManager}" is not supported.`);

				return undefined;
			}

			const isWindows = os.platform() === 'win32';
			const globalPath = await resolver(trace, isWindows);

			if (globalPath) {
				cache[packageManager] = globalPath;
			}

			return globalPath;
		},
	};
}

module.exports = {
	getGlobalPathResolver,
};
