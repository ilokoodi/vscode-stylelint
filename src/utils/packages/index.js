'use strict';

module.exports = {
	...require('./find-package-root'),
	...require('./global-path-resolver'),
	...require('./stylelint-resolver'),
};
