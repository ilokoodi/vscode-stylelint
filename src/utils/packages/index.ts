export * from './find-package-root';
export * from './global-path-resolver';
export * from './stylelint-resolver';
export * from './types';
