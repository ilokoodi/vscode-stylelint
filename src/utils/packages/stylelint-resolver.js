'use strict';

const path = require('path');
const { Files } = require('vscode-languageserver/node');
const { URI } = require('vscode-uri');
const { getWorkspaceFolder } = require('../documents');
const { getGlobalPathResolver } = require('./global-path-resolver');

/**
 * Utility for resolving the path to the Stylelint package. Each instance caches
 * resolved paths to global `node_modules` directories.
 */
class StylelintResolver {
	/**
	 * @param {lsp.Connection} [connection] The language server connection.
	 */
	constructor(connection) {
		this._globalPathResolver = getGlobalPathResolver();
		this._connection = connection;
	}

	/**
	 * Logs an error message through the connection if one is available.
	 * @private
	 * @param {string} message The message to log.
	 * @param {boolean} showMessage Whether to show the message to the user in a notification.
	 * @returns {void}
	 */
	_logError(message, showMessage = true) {
		if (!this._connection) {
			return;
		}

		if (showMessage) {
			this._connection.window.showErrorMessage(`stylelint: ${message}`);
		}

		this._connection.console.error(message);
	}

	/**
	 * Attempts to resolve the stylelint package from a path. If an error
	 * occurs, it will be logged through the connection and thrown. If the
	 * resolved module does not have a lint function, an error will be logged
	 * and `undefined` will be returned.
	 * @private
	 * @param {string} stylelintPath
	 * @returns {stylelint.PublicApi | undefined}
	 */
	_resolveFromPath(stylelintPath) {
		const errorMessage = `Failed to load stylelint from "stylelintPath": ${stylelintPath}.`;

		let stylelint;

		try {
			stylelint = require(stylelintPath);
		} catch (err) {
			this._logError(errorMessage);
			throw err;
		}

		if (stylelint && typeof stylelint.lint === 'function') {
			return stylelint;
		}

		this._logError(errorMessage);

		return undefined;
	}

	/**
	 * Attempts to resolve the stylelint package from the given document's
	 * workspace folder or the global `node_modules` directory for the given
	 * package manager. Resolution will be traced through the connection.
	 *
	 * If a path cannot be resolved, `undefined` will be returned. If the
	 * resolved module does not have a lint function, an error will be logged
	 * and `undefined` will be returned.
	 * @private
	 * @param {lsp.TextDocument} textDocument
	 * @param {PackageManager} [packageManager]
	 * @returns {Promise<stylelint.PublicApi | undefined>}
	 */
	async _resolveFromModules(textDocument, packageManager) {
		const connection = this._connection;

		/** @type {TracerFn} */
		const trace = connection
			? (message, verbose) => {
					connection.tracer.log(message, verbose);
			  }
			: () => undefined;

		/** @type {stylelint.PublicApi | undefined} */
		let stylelint;

		try {
			/** @type {string | undefined} */
			const globalModulesPath = packageManager
				? await this._globalPathResolver.resolve(packageManager, trace)
				: undefined;

			const documentURI = URI.parse(textDocument.uri);

			const cwd =
				documentURI.scheme === 'file'
					? path.dirname(documentURI.fsPath)
					: await getWorkspaceFolder(textDocument, connection);

			const stylelintPath = await Files.resolve('stylelint', globalModulesPath, cwd, trace);

			stylelint = require(stylelintPath);

			if (stylelint && typeof stylelint.lint !== 'function') {
				this._logError('stylelint.lint is not a function.');

				return undefined;
			}
		} catch {
			// ignore
		}

		return stylelint;
	}

	/**
	 * Attempts to resolve the `stylelint` package from the following lcoations,
	 * in order:
	 *
	 * 1. `options.stylelintPath`, if provided.
	 * 2. `node_modules` in the workspace folder of the given document.
	 * 3. The global `node_modules` directory for the given package manager.
	 *
	 * If `options.stylelintPath` is provided, but the path to which it points
	 * cannot be required, an error will be thrown. In all other cases of failed
	 * resolution, `undefined` will be returned. Resolution fails if either the
	 * path to the `stylelint` package cannot be resolved or if the resolved
	 * module does not have a `lint` function.
	 *
	 * If a connection is available, errors will be logged through it and module
	 * resolution through `node_modules` will be traced through it.
	 * @param {StylelintVSCodeOptions} options
	 * @param {lsp.TextDocument} textDocument
	 * @returns {Promise<stylelint.PublicApi | undefined>}
	 */
	async resolve({ packageManager, stylelintPath }, textDocument) {
		const stylelint =
			(stylelintPath ? this._resolveFromPath(stylelintPath) : null) ??
			(await this._resolveFromModules(textDocument, packageManager));

		if (!stylelint) {
			this._logError(
				'Failed to load stylelint either globally or from the current workspace.',
				false,
			);

			return undefined;
		}

		return stylelint;
	}
}

module.exports = {
	StylelintResolver,
};
