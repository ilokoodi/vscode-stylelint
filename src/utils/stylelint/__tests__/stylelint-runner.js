'use strict';

jest.mock('os');
jest.mock('vscode-uri', () => ({
	URI: {
		parse: jest.fn((/** @type {string} */ str) => ({ fsPath: str })),
	},
}));
jest.mock('../../packages/stylelint-resolver');

const stylelint = require('stylelint');

const mockedOS = /** @type {tests.mocks.OSModule} */ (require('os'));
const resolver = /** @type {jest.Mocked<typeof import('../../packages/stylelint-resolver')>} */ (
	require('../../packages/stylelint-resolver')
);

const { StylelintRunner } = require('../stylelint-runner');

/** @typedef {(options: stylelint.LinterOptions) => Promise<Partial<stylelint.LinterResult>>} FakeLintFunction */

/**
 * @param {FakeLintFunction} [lint]
 * @param {(serverOptions: StylelintVSCodeOptions, document: lsp.TextDocument) => Promise<{lint: FakeLintFunction}>} [resolve]
 * @returns {() => import('../../packages/stylelint-resolver').StylelintResolver}
 */
const createMockResolver = (lint, resolve) =>
	/** @type {any} */ (
		() => ({
			resolve: resolve ?? (async () => (lint ? { lint } : undefined)),
		})
	);

/**
 * @param {string} code
 * @param {string} [uri]
 * @returns {lsp.TextDocument}
 */
const createMockDocument = (code, uri = '/path/to/file.css') =>
	/** @type {lsp.TextDocument} */ ({
		getText: () => code,
		uri,
	});

/** @type {lsp.Connection} */
const mockConnection = /** @type {any} */ ({});

describe('StylelintRunner', () => {
	beforeEach(() => {
		mockedOS.__mockPlatform('linux');
	});

	test('should return no diagnostics if Stylelint cannot be resolved', async () => {
		resolver.StylelintResolver.mockImplementation(createMockResolver());

		const results = await new StylelintRunner(mockConnection).lintDocument(createMockDocument(''));

		expect(results).toEqual({ diagnostics: [] });
	});

	// TODO: Remove once fixed upstream
	test('should upper-case drive letters on Windows (Stylelint bug #5594)', async () => {
		expect.assertions(2);

		mockedOS.__mockPlatform('win32');
		resolver.StylelintResolver.mockImplementation(
			createMockResolver(async (options) => {
				expect(options.codeFilename).toBe('C:\\path\\to\\file.css');

				return { results: [] };
			}),
		);

		await new StylelintRunner(mockConnection).lintDocument(
			createMockDocument('', 'c:\\path\\to\\file.css'),
		);

		mockedOS.__mockPlatform('linux');
		resolver.StylelintResolver.mockImplementation(
			createMockResolver(async (options) => {
				expect(options.codeFilename).toBe('c:/path/to/file.css');

				return { results: [] };
			}),
		);

		await new StylelintRunner(mockConnection).lintDocument(
			createMockDocument('', 'c:/path/to/file.css'),
		);
	});

	test('should call stylelint.lint with the document path and given options', async () => {
		expect.assertions(2);

		resolver.StylelintResolver.mockImplementation(
			createMockResolver(async (options) => {
				expect(options).toEqual({
					code: 'a {}',
					codeFilename: '/path/to/file.scss',
					config: {
						customSyntax: 'postcss-scss',
					},
					fix: true,
					formatter: expect.any(Function),
				});

				expect(/** @type {stylelint.Formatter} */ (options.formatter)?.([])).toBe('');

				return { results: [] };
			}),
		);

		await new StylelintRunner(mockConnection).lintDocument(
			createMockDocument('a {}', '/path/to/file.scss'),
			{
				config: {
					customSyntax: 'postcss-scss',
				},
				fix: true,
			},
		);
	});

	test("should pass empty rules if the document's path cannot be determined and rules aren't set", async () => {
		expect.assertions(1);

		resolver.StylelintResolver.mockImplementation(
			createMockResolver(async (options) => {
				expect(options.config).toEqual({
					rules: {},
				});

				return { results: [] };
			}),
		);

		await new StylelintRunner(mockConnection).lintDocument(createMockDocument('a {}', ''));
	});

	test("should not change set rules if the document's path cannot be determined", async () => {
		expect.assertions(1);

		resolver.StylelintResolver.mockImplementation(
			createMockResolver(async (options) => {
				expect(options.config).toEqual({
					rules: { 'no-empty-source': true },
				});

				return { results: [] };
			}),
		);

		await new StylelintRunner(mockConnection).lintDocument(createMockDocument('a {}', ''), {
			config: { rules: { 'no-empty-source': true } },
		});
	});

	test('should call the resolver with the server options and the document', async () => {
		expect.assertions(3);

		resolver.StylelintResolver.mockImplementation(
			createMockResolver(undefined, async (serverOptions, document) => {
				expect(serverOptions).toEqual({
					packageManager: 'pnpm',
				});

				expect(document).toEqual({
					getText: expect.any(Function),
					uri: '/path/to/file.css',
				});

				expect(document.getText()).toBe('a {}');

				return {
					lint: async () => ({ results: [] }),
				};
			}),
		);

		await new StylelintRunner(mockConnection).lintDocument(
			createMockDocument('a {}', '/path/to/file.css'),
			undefined,
			{ packageManager: 'pnpm' },
		);
	});

	test('should return processed lint results from Stylelint without configured rules', async () => {
		expect.assertions(1);

		resolver.StylelintResolver.mockImplementation(
			createMockResolver(undefined, async () => stylelint),
		);

		const results = await new StylelintRunner(mockConnection).lintDocument(
			createMockDocument('table {', '/path/to/file.css'),
		);

		expect(results).toMatchSnapshot();
	});

	test('should return processed lint results from Stylelint with configured rules', async () => {
		expect.assertions(1);

		resolver.StylelintResolver.mockImplementation(
			createMockResolver(undefined, async () => stylelint),
		);

		const results = await new StylelintRunner(mockConnection).lintDocument(
			createMockDocument('a {}', '/path/to/file.css'),
			{ config: { rules: { 'block-no-empty': true } } },
		);

		expect(results).toMatchSnapshot();
	});

	test('should throw errors thrown by Stylelint', async () => {
		expect.assertions(1);

		resolver.StylelintResolver.mockImplementation(
			createMockResolver(undefined, async () => stylelint),
		);

		await expect(
			new StylelintRunner(mockConnection).lintDocument(
				createMockDocument('a {}', '/path/to/file.css'),
				{
					config: { rules: {} },
					files: ['/path/to/file.css'],
				},
			),
		).rejects.toThrowErrorMatchingSnapshot();
	});
});
