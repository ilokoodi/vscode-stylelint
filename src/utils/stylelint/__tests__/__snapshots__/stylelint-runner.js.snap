// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StylelintRunner should return processed lint results from Stylelint with configured rules 1`] = `
Object {
  "diagnostics": Array [
    Object {
      "code": "block-no-empty",
      "codeDescription": Object {
        "href": "https://stylelint.io/user-guide/rules/block-no-empty",
      },
      "message": "Unexpected empty block (block-no-empty)",
      "range": Object {
        "end": Object {
          "character": 2,
          "line": 0,
        },
        "start": Object {
          "character": 2,
          "line": 0,
        },
      },
      "severity": 1,
      "source": "stylelint",
    },
  ],
}
`;

exports[`StylelintRunner should return processed lint results from Stylelint without configured rules 1`] = `
Object {
  "diagnostics": Array [
    Object {
      "code": "CssSyntaxError",
      "message": "Unclosed block (CssSyntaxError)",
      "range": Object {
        "end": Object {
          "character": 0,
          "line": 0,
        },
        "start": Object {
          "character": 0,
          "line": 0,
        },
      },
      "severity": 1,
      "source": "stylelint",
    },
  ],
}
`;

exports[`StylelintRunner should throw errors thrown by Stylelint 1`] = `"You must pass stylelint a \`files\` glob or a \`code\` string, though not both"`;
